apps/aecore/src/aec_conductor.erl

```diff
@@ -0,0 +1,601 @@
+    retry_mining(Candidate, State).
+
+%% NOTE: State is passed through in preparation for a functional chain State.
+save_mined_block(Block, State) ->
+    Header = aec_blocks:to_header(Block),
+    case aec_chain:insert_header(Header) of
+        ok ->
+            case aec_chain:write_block(Block) of
+                ok ->
+                    try exometer:update([ae,epoch,aecore,mining,blocks_mined], 1)
+                    catch error:_ -> ok end,
+                    epoch_mining:info("Block inserted: Height = ~p"
+                                      "~nHash = ~s",
+                                      [Block#block.height,
+                                       as_hex(Block#block.root_hash)]),
+                    aec_events:publish(block_created, Block),
+                    State;
+                {error, Reason} ->
+                    epoch_mining:error("Block insertion failed: ~p.", [Reason]),
+                    State
+            end;
+        {error, Reason} ->
+            epoch_mining:error("Header insertion failed: ~p.", [Reason]),
+            State
+    end.
+
+as_hex(S) ->
+    [io_lib:format("~2.16.0b",[X]) || <<X:8>> <= S].
+
+%%%===================================================================
+%%% Retry mining when we failed to find a solution.
+
+retry_mining(#candidate{top_hash = Hash1},
+             #state{seen_top_block_hash = Hash2} =State) when Hash1 =/= Hash2 ->
+    %% Stale hash for candidate. Rebuild it.
+    start_mining(State#state{block_candidate = undefined});
+retry_mining(Candidate, #state{fetch_new_txs_from_pool = true} = State) ->
+    Block = Candidate#candidate.block,
+    %% We should first see if we can get a new candidate.
+    case aec_mining:apply_new_txs(Block) of
+        {ok, Block} ->
+            epoch_mining:info("No new txs available"),
+            retry_mining_with_new_nonce(Candidate, State);
+        {ok, NewBlock, RandomNonce} ->
+            epoch_mining:info("New txs added for mining"),
+            NewNonce = aec_pow:next_nonce(RandomNonce),
+            NewCandidate = new_candidate(NewBlock, NewNonce, RandomNonce, State),
+            start_mining(State#state{block_candidate = NewCandidate});
+        {error, What} ->
+            epoch_mining:error("Retrying with new nonce after unexpected error"
+                               " in applying new txs: ~p", [What]),
+            retry_mining_with_new_nonce(Candidate, State)
+    end;
+retry_mining(Candidate, State) ->
+    retry_mining_with_new_nonce(Candidate, State).
+
+retry_mining_with_new_nonce(#candidate{nonce = N, max_nonce = N}, State) ->
+    epoch_mining:info("Failed to mine block, "
+                      "nonce range exhausted (was ~p)", [N]),
+    start_mining(State#state{block_candidate = undefined});
+retry_mining_with_new_nonce(Candidate, State) ->
+    Nonce = Candidate#candidate.nonce,
+    NewNonce = aec_pow:next_nonce(Nonce),
+    epoch_mining:info("Not fetching new txs; "
```

Comment:
Should be a debug message, or perhaps a metric update

Reply:
I'll change it